### ISSUE TITLE
0.9.9.8 Linux build feedback and fixes

### ISSUE THREAD

**Issue by tung** [chaosforgeorg/doomrl#141]
I'm on Debian 12 and I can confirm that the game compiles and runs, but only after a couple of source code tweaks.

Steps:

1. Install dependencies: `sudo install lua5.1 liblua5.1-0-dev fp-compiler fp-units-base fp-units-fcl fp-units-misc`
2. Also install SDL 2 libs, if you don't already have them: `sudo install libsdl2-2.0-0 libsdl2-image-2.0-0 libsdl2-mixer-2.0-0 libsdl2-ttf-2.0-0`
3. Get the DRL and FPC Valkyrie source code and put their directories are next to each other.
4. Get the game download from <https://drl.chaosforge.org> and extract the files from `mp3`, `music`, `wav` and `wavhq` into the matching sub-directories in DRL's `bin` directory.
5. In the DRL directory, `mkdir tmp` otherwise compilation fails with the following message:

```
vutil.pas(800,1) Error: Can't create object file: tmp/vutil.o (error code: 2)
vutil.pas(800,1) Fatal: Can't create object tmp/vutil.o
```

6. In the FPC Valkyrie code, comment out references to `IMG_LoadWEBPAnimation_RW`; Debian 12's SDL2_Image doesn't seem to have it and DRL doesn't need it.

```diff
diff --git a/libs/vsdl2imagelibrary.pas b/libs/vsdl2imagelibrary.pas
index 0808885..fa497f5 100644
--- a/libs/vsdl2imagelibrary.pas
+++ b/libs/vsdl2imagelibrary.pas
@@ -103,7 +103,7 @@ var
   IMG_LoadAnimationTyped_RW       : function (src: PSDL_RWops; freesrc: Integer; const _type: PChar): PIMG_Animation; cdecl;
   IMG_FreeAnimation               : procedure (anim: PIMG_Animation); cdecl;
   IMG_LoadGIFAnimation_RW         : function (src: PSDL_RWops): PIMG_Animation; cdecl;
-  IMG_LoadWEBPAnimation_RW        : function (src: PSDL_RWops): PIMG_Animation; cdecl;
+  {IMG_LoadWEBPAnimation_RW        : function (src: PSDL_RWops): PIMG_Animation; cdecl;}
 
 var
   SDL2_image : TLibrary = nil;
@@ -184,7 +184,7 @@ begin
   Pointer(IMG_LoadAnimationTyped_RW) := GetSymbol('IMG_LoadAnimationTyped_RW');
   Pointer(IMG_FreeAnimation) := GetSymbol('IMG_FreeAnimation');
   Pointer(IMG_LoadGIFAnimation_RW) := GetSymbol('IMG_LoadGIFAnimation_RW');
-  Pointer(IMG_LoadWEBPAnimation_RW) := GetSymbol('IMG_LoadWEBPAnimation_RW');
+  {Pointer(IMG_LoadWEBPAnimation_RW) := GetSymbol('IMG_LoadWEBPAnimation_RW');}
 
   Exit( True );
 end;
```
7. Run the build: `lua5.1 makefile.lua`
8. Run the game: `cd bin` and `./drl`

In this state, the game makes a window but crashes immediately; seems like there's a problem with `FLayers` being nil in `TDoomIO.Destroy` in `src/doomio.pas`, so patch it like this and build it again:

```diff
diff --git a/src/doomio.pas b/src/doomio.pas
index c9a5465..4e1d945 100644
--- a/src/doomio.pas
+++ b/src/doomio.pas
@@ -541,8 +541,9 @@ begin
   FreeAndNil( FOldASCII );
   FreeAndNil( FNewASCII );
 
-  for iLayer in FLayers do
-    iLayer.Free;
+  if FLayers <> nil then
+    for iLayer in FLayers do
+      iLayer.Free;
   FreeAndNil( FLayers );
   IO := nil;
   inherited Destroy;
```

At this point the game should launch and run successfully.

I think you can type just `sudo lua5.1 liblua5.1-0-dev fpc`  instead of all the `fp-whatever` stuff at the first step, but it'll install a lot more than what you strictly need to build the game.

**Comment by epyon** [comment]
Thanks! I'll integrate it to Monday's patch and try to build official Linux builds (I failed to notice a few things you mentioned here and was running against the clock at release day).

**Comment by tung** [comment]
No worries, I understand how busy release days can be; just thought you might appreciate the testing.

**Comment by dderevjanik** [comment]
Do you think it would be possible to use Docker in order to avoid installing dependencies? 

@epyon do you have any plan to release on mac with ARM?

**Comment by epyon** [comment]
If someone supplies a mac build/patch, I'd gladly integrate it and put it up as official. Myself I no longer have access to any mac's at all :/

**Comment by dderevjanik** [comment]
Thank you for your response. Setting up the development environment on my Mac has been quite challenging, which is why I was asking about using Docker.

**Comment by epyon** [comment]
Linux binaries are up! Please test them if all is ok, and I'll close this issue!

**Comment by dderevjanik** [comment]
Do you have any plans to release OSX binaries as well?

Additionally, I have created a Dockerfile for compiling DRL, but I'm not sure if anything is missing. Cannot find `drl` in `doomrl/bin`. What do you think @tung ?

```dockerfile
# Use an official Ubuntu as a base image
FROM ubuntu:22.04

WORKDIR /chaosforgeorg
# Install necessary packages
RUN apt update
RUN apt install -y \
    git \
    curl \
    unzip
RUN apt install -y \
    lua5.1 \
    liblua5.1-0-dev \
    fp-compiler \
    fp-units-base \
    fp-units-fcl \
    fp-units-misc
RUN apt install -y \
    libsdl2-2.0-0 \
    libsdl2-image-2.0-0 \
    libsdl2-mixer-2.0-0 \
    libsdl2-ttf-2.0-0
RUN git clone https://github.com/chaosforgeorg/doomrl.git \
    && git clone https://github.com/chaosforgeorg/fpcvalkyrie.git
RUN curl -L -o drl-win-0998.zip https://drl.chaosforge.org/file_download/41/drl-win-0998.zip
RUN unzip drl-win-0998.zip \
    && cp drl-win-0998/mp3/* doomrl/bin/mp3 \
    && cp drl-win-0998/wavhq/* doomrl/bin/wavhq
ENV HOME=/home
RUN cd doomrl && lua5.1 ./makefile.lua
```

**Comment by tung** [comment]
The Linux binaries on the website need glibc 2.38, which means they're built for either Ubuntu 23.10/24.04 or Debian unstable/testing, and I only have a Debian 12 stable setup here, so someone else will have to test them.  I can at least confirm that the current versions of DRL and FPC Valkyrie compile with no changes, except for having to make the "tmp" directory before building.

@dderevjanik Make the "tmp" directory before building; it might also be a good idea to check out specific tags/commits of the drl and fpcvalkyrie repos if you want it to be reproducible.

**Comment by dderevjanik** [comment]
Thanks! I tried it with `mkdir tmp`, but another error occurs. Have no idea what's going on.. 

```
#13 [10/10] RUN cd doomrl && mkdir tmp && echo 'done' && lua5.1 ./makefile.lua
#13 13.49 done
#13 17.83 Free Pascal Compiler version 3.2.2+dfsg-9ubuntu1 [2022/04/11] for aarch64
#13 17.83 Copyright (c) 1993-2021 by Florian Klaempfl and others
#13 20.53 vgenerics.pas(183,26) Hint: Parameter "aItem" not used
#13 20.53 vgenerics.pas(271,26) Hint: Parameter "aItem" not used
#13 21.61 vluaext.pas(484,49) Hint: Local variable "lnumber" does not seem to be initialized
#13 32.13 vfmod2library.pas(334,3) Error: Identifier not found "Set8087CW"
#13 32.13 vfmod2library.pas(706,5) Error: Identifier not found "Set8087CW"
#13 32.13 vfmod2library.pas(711) Fatal: There were 2 errors compiling module, stopping
#13 32.13 Fatal: Compilation aborted
#13 32.13 Error: /usr/bin/ppca64 returned an error exitcode
#13 32.14 Autodetected OS - LINUX
#13 DONE 32.4s
```

**Comment by draeath** [comment]
This doesn't want to run on OpenSUSE Tumbleweed:

    draeath@ginnungagap:/opt/draeath/drl-linux-0998> ldd drl
    ./drl: /lib64/libncursesw.so.6: version `NCURSESW6_5.6.20061217' not found (required by ./drl)
    ./drl: /lib64/libncursesw.so.6: version `NCURSESW6_5.3.20021019' not found (required by ./drl)
    ./drl: /lib64/libncursesw.so.6: version `NCURSESW6_5.1.20000708' not found (required by ./drl)
            linux-vdso.so.1 (0x00007ffc32f6d000)
            libm.so.6 => /lib64/libm.so.6 (0x00007ff515cdc000)
            libX11.so.6 => /lib64/libX11.so.6 (0x00007ff515b95000)
            libncursesw.so.6 => /lib64/libncursesw.so.6 (0x00007ff515b55000)
            libtinfo.so.6 => /lib64/libtinfo.so.6 (0x00007ff515b18000)
            libc.so.6 => /lib64/libc.so.6 (0x00007ff515800000)
            /lib64/ld-linux-x86-64.so.2 (0x00007ff515dea000)
            libxcb.so.1 => /lib64/libxcb.so.1 (0x00007ff515aec000)
            libXau.so.6 => /lib64/libXau.so.6 (0x00007ff515ae7000)
    draeath@ginnungagap:~> rpm -qa | grep libncurses6
    libncurses6-6.5.20240817-43.1.x86_64
    libncurses6-32bit-6.5.20240817-43.1.x86_64

I tried (and completely failed) to build it myself, the furthest I could get was Lazarus whining about undefined symbols (I didn't keep the details).

**Comment by jason-mehmel** [comment]
> The Linux binaries on the website need glibc 2.38, which means they're built for either Ubuntu 23.10/24.04 or Debian unstable/testing, and I only have a Debian 12 stable setup here, so someone else will have to test them. I can at least confirm that the current versions of DRL and FPC Valkyrie compile with no changes, except for having to make the "tmp" directory before building.
> 
> @dderevjanik Make the "tmp" directory before building; it might also be a good idea to check out specific tags/commits of the drl and fpcvalkyrie repos if you want it to be reproducible.

I had the same problem re: glibc 2.38 when trying to run the version for download on the website. Haven't tried to compile it myself yet.

If I tried to compile it, would that then end up using glibc 2.36 (which is in Stable) or is that dependency based on the calls inside the code?

**Comment by dderevjanik** [comment]
@jason-mehmel thanks!

I was able to fix that problem by installing `libx11-dev` and running `ldconfig` before compiling drl. There was another issue, that I used `drl.wad` from Windows build instead of Linux one (how to build drl.wad @epyon ?) .

Now, I'm facing another issue 

```
Abnormal program termination!

Reason : Can't load library "libSDL2_image-2.0.so.0": libSDL2_image-2.0.so.0: cannot open shared object file: No such file or directory

If this reason doesn't seem your fault, please submit a bug report at
http://forum.chaosforge.org/, be sure to include the last entries in
your error.log that will get created once you hit Enter.
```

Any ideas how to fix it?

If anyone is interested, here's my latest Dockerfile with compile script https://gist.github.com/dderevjanik/1be7840284b58af65a32ce2a7d15a840

**Comment by epyon** [comment]
Are the compilation issues sorted out?

As far as the build goes, I've set up a new build environment that should have weaker reqs, if anyone could check the BETA 3 that has just been posted, I'd be glad : https://forum.chaosforge.org/index.php?topic=8893.msg71140#msg71140

**Comment by draeath** [comment]
@epyon Permissions in the tarball distribution are borked (everything is 0755). That aside: no joy on OpenSUSE:

    draeath@muspelheimr:~/Downloads/drl-linux-0999-beta3-lq> ./drl
    ./drl: /lib64/libncursesw.so.6: version `NCURSESW6_5.6.20061217' not found (required by ./drl)
    ./drl: /lib64/libncursesw.so.6: version `NCURSESW6_5.3.20021019' not found (required by ./drl)
    ./drl: /lib64/libncursesw.so.6: version `NCURSESW6_5.1.20000708' not found (required by ./drl)
    draeath@muspelheimr:~/Downloads/drl-linux-0999-beta3-lq> ldd drl
    ./drl: /lib64/libncursesw.so.6: version `NCURSESW6_5.6.20061217' not found (required by ./drl)
    ./drl: /lib64/libncursesw.so.6: version `NCURSESW6_5.3.20021019' not found (required by ./drl)
    ./drl: /lib64/libncursesw.so.6: version `NCURSESW6_5.1.20000708' not found (required by ./drl)
            linux-vdso.so.1 (0x00007ffc401c3000)
            libm.so.6 => /lib64/libm.so.6 (0x00007fe111fbd000)
            libX11.so.6 => /lib64/libX11.so.6 (0x00007fe111e74000)
            libncursesw.so.6 => /lib64/libncursesw.so.6 (0x00007fe111e34000)
            libtinfo.so.6 => /lib64/libtinfo.so.6 (0x00007fe111df6000)
            libc.so.6 => /lib64/libc.so.6 (0x00007fe111a00000)
            /lib64/ld-linux-x86-64.so.2 (0x00007fe1120d6000)
            libxcb.so.1 => /lib64/libxcb.so.1 (0x00007fe111dca000)
            libXau.so.6 => /lib64/libXau.so.6 (0x00007fe111dc5000)
    draeath@muspelheimr:~/Downloads/drl-linux-0999-beta3-lq> zypper --no-refresh info libncurses6
    Loading repository data...
    Reading installed packages...
    
    
    Information for package libncurses6:
    ------------------------------------
    Repository     : Main Repository (OSS)
    Name           : libncurses6
    Version        : 6.5.20250104-52.1
    Arch           : x86_64
    Vendor         : openSUSE
    Installed Size : 1.2 MiB
    Installed      : Yes (automatically)
    Status         : up-to-date
    Source package : ncurses-6.5.20250104-52.1.src
    Upstream URL   : https://www.invisible-island.net/ncurses/ncurses.html
    Summary        : Terminal control library
    Description    : 
        The ncurses library is used by many terminal applications for
        controlling output to the screen and input from the user.
        
        This package contains the library built with the version 6 ABI.

It still wants a (far too) specific version of the ncurses6 library.

I don't know if the build process has improved at all, but I figured I could at least try the BETA3 build for you.

**Comment by epyon** [comment]
Will take a look at ncurses. What should the permissions be, apart from +x for executables? (sorry I don't use nixes at all except a WSL shell)

**Comment by draeath** [comment]
Hey, no worries! You only know what you've practiced!

0644 (rw,r,r) for regular files, 0755 (rwx,rx,rx) for executables and directories are sane permissions.

I wish I could help re: the library versioning. I don't know what determines how specific the linker wants to be, but usually any random make or cmake project doesn't seem to do that by default.

**Comment by epyon** [comment]
Unfortunately it's freepascal's ncurses unit that decides that. I'll try building a version Monday that statically links ncurses.

**Comment by epyon** [comment]
Shit, I built on the wrong image. I'll upload a new version Monday.

**Comment by epyon** [comment]
Scratch Monday - I've rebuilt the binaries with statically linked ncurses and on Ubuntu 20.04
https://www.dropbox.com/scl/fi/hdpoyurv254jhub2asheb/drl-linux-0999-beta3-lq-2.tar.gz?rlkey=df4rtw1bxf91bx5i7rkuxh58p&dl=0?rlkey=0wp080tvdets4ng2mdfl18abe&dl=0

**Comment by jason-mehmel** [comment]
I can't speak for Ubuntu, but on my machine it works fine!

Distro: BunsenLabs GNU/Linux 12 (Boron) base: Debian GNU/Linux 12


~Jason
pronouns: he / him / his
***@***.*** | www.jasonmehmel.com

**Comment by draeath** [comment]
I'm still seeing it trying to link to `libncursesw.so.6` and getting mad about the version.

Just to make sure I got the updated build, does this look right?

    draeath@muspelheimr:~/Downloads/drl-linux-0999-beta3-lq> md5sum drl
    4bd64ab39cd9ae0840fe13bb07d97d0b  drl

**Comment by draeath** [comment]
What's really fun, is if I take `libncurses6_6.2-0ubuntu2.1_amd64.deb` and `libncursesw6_6.2-0ubuntu2.1_amd64.deb` from the Ubuntu 20.04 archive and slap that all next to DRL and try to run with `LD_LIBRARY_PATH=.` - even though `ld` is happy, the game fails with another symbol issue:

    draeath@muspelheimr:~/Downloads/drl-linux-0999-beta3-lq> LD_LIBRARY_PATH=. ldd drl
            linux-vdso.so.1 (0x00007ffe48fe2000)
            libm.so.6 => /lib64/libm.so.6 (0x00007f06d686d000)
            libdl.so.2 => /lib64/libdl.so.2 (0x00007f06d6868000)
            libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f06d6863000)
            libX11.so.6 => /lib64/libX11.so.6 (0x00007f06d6718000)
            libncursesw.so.6 => ./libncursesw.so.6 (0x00007f06d66de000)
            libtinfo.so.6 => /lib64/libtinfo.so.6 (0x00007f06d66a2000)
            libc.so.6 => /lib64/libc.so.6 (0x00007f06d6400000)
            /lib64/ld-linux-x86-64.so.2 (0x00007f06d6986000)
            libxcb.so.1 => /lib64/libxcb.so.1 (0x00007f06d6676000)
            libXau.so.6 => /lib64/libXau.so.6 (0x00007f06d666f000)
    draeath@muspelheimr:~/Downloads/drl-linux-0999-beta3-lq> LD_LIBRARY_PATH=. ./drl
    ./drl: symbol lookup error: ./libncursesw.so.6: undefined symbol: TABSIZE, version NCURSES6_TINFO_5.0.19991023
    draeath@muspelheimr:~/Downloads/drl-linux-0999-beta3-lq>

**Comment by dderevjanik** [comment]
Guys, have you considered creating a Dockerfile for this project? It could address several challenges related to setting up the development and build environments while also providing additional benefits, such as:

- Quickly spin up a consistent development environment for all contributors.
- The Dockerfile can act as a clear, self-contained reference for configuring the environment.
- Integrate with GitHub Actions for tasks like nightly builds or releases triggered on every commit.

**Comment by epyon** [comment]
The latest version:
https://forum.chaosforge.org/index.php/topic,8895.msg71144/topicseen.html#msg71144

What has been done:
- added the OP source file suggestions
- reduced build Ubuntu environment to 20.04 which should relieve the high glibc requirement
- build system tries to set proper permissions on the files
- build system tries to set proper line endings for script files

What has not been done:
- nothing has been done about the ncurses version - FPC links to that one and there's no way to convince him otherwise - static linking opens a huge can of worms, patches welcome if someone knows how to fix this
- docker files are beyond my scope, if someone wants to do that feel free!

I'll be closing this issue unless there's something to add to the previous fixes- feel free to open a new issue related to the two "not done" above if you have any suggestions how to adress that.
